tree/ntuple/v7/doc/architecture.md

```diff
@@ -3,19 +3,468 @@ RNTuple Code Architecture
+
+A model can add _projected fields_.
+Projected fields map existing physical fields to a different type.
+For instance, a `std::vector<Event>` can be projected onto a `std::vector<float>` for a float member of `Event`.
+Projected fields are stored as header meta-data.
+
+Fields can be added to a model after the writing process has started (cf. `RNTupleWriter::CreateModelUpdater()`).
+This is called _late model extension_.
+Addition of fields invalidates previously created entries.
+The values of deferred fields for the already written entries is set to the default constructed type of the field.
+
+### REntry
+The REntry represents a row/entry in an RNTuple.
+It contains a list of `RValue` objects that correspond to the top-level fields of the originating model.
+The entry gives access to the shared pointers corresponding to the top-level fields.
+It also provides functionality to bind application-provided pointers.
+
+An REntry can be passed to `RNTupleWriter::Fill()` and `RNTupleReader::LoadEntry()`.
+Otherwise, the reader/writer uses the default entry of its model.
+
+An entry can safely outlive its originating model.
+New objects cannot anymore be created (`EmplaceNewValue` will throw an exception), but the entry is still properly destructed.
+
+### RNTupleWriter, RNTupleParallelWriter
+The RNTupleWriter is the primary interface to create an RNTuple.
+The writer takes ownership of a given model.
+The writer can either add an RNTuple to an existing ROOT file (`RNTupleWriter::Append()`) or create a new ROOT file with an RNTuple (`RNTupleWriter::Recreate()`).
+Once created, entries are added to an RNTuple either serially (`RNTupleWriter::Fill()`) or in concurrently in multiple threads with the `RNTupleParallelWriter`.
+Once committed (e.g. by releasing the RNTupleWriter), the RNTuple is immutable and cannot be ammended.
+An RNTuple that is currently being written cannot be read.
+
+### RNTupleReader
+The RNTupleReader is the primary interface to read and inspect an RNTuple.
+An RNTupleReader owns a model: either a model created from the on-disk information or an imposed, user-provided model.
+The user-provided model can be limited to a subset of fields.
+Data is populated to an explicit `REntry` or the model's default entry through `RNTupleReader::LoadEntry()`.
+
+The reader can create `RNTupleView` objects for the independent reading of individual fields.
+The reader can create `RBulk` objects for bulk reading of individual fields.
+
+Additionally, the reader provides access to a cached copy of the descriptor.
+It can display individual entries (`RNTupleReader::Show()`) and summary information (`RNTupleReader::PrintInfo()`).
+
+### RNTupleView<T>
+RNTuple views provide read access to individual fields.
+Views are created from an RNTupleReader.
+Views are templated; for simple types (e.g., `float`, `int`), views provide read-only access directly to an RNTuple page in memory.
```

Comment:
Unless the user provides a memory location for it after https://github.com/root-project/root/pull/14849 . But we can amend the document if/when that PR is merged